lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,355 @@
+        divergentUses.insert(cast<Instruction>(use.get()));
+    }
+  }
+
+  // Stop propagation due to violating the second condition
+  divergentValues.insert(firstDivergentValue);
+  for (auto use : divergentUses) {
+    auto findIter = divergentValues.find(use);
+    if (findIter == divergentValues.end()) {
+      m_candidateInsts.insert(inst);
+      return;
+    }
+  }
+  divergentValues.erase(firstDivergentValue);
+
+  // Take action on each operands of inst
+  for (auto value : divergentValues)
+    takeAction(value);
+}
+
+// =====================================================================================================================
+// Take action on the given instruction
+//
+// @param inst : The instruction used to descide the next step
+void PatchReadFirstLane::takeAction(Instruction *inst) {
+  // No action on the already processed instruction
+  if (m_markedUniforms.find(inst) != m_markedUniforms.end())
+    return;
+  // Not ready for a decision
+  if (!isAllUsersMarkedUniform(inst))
+    return;
+
+  // Stop propagation when the instruction for which isSourceOfDivergence is true
+  // NOTE: Stop propagation for a loadInst in case it is fat pointer due to uncertain size of a fat pointer
+  bool stopPropagation = false;
+  if (m_targetTransformInfo)
+    stopPropagation = m_targetTransformInfo->isSourceOfDivergence(inst);
+  else
+    stopPropagation = dyn_cast<CallInst>(inst);
+
+  if (!stopPropagation) {
+    if (auto loadInst = dyn_cast<LoadInst>(inst))
+      stopPropagation = (loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER);
+  }
+
+  // Take action by stopping propagation or continue propagation
+  if (stopPropagation)
+    m_candidateInsts.insert(inst);
+  else
+    m_markedUniforms.insert(inst);
+}
+
+// =====================================================================================================================
+// Check whether all users of the given instruction are uniform
+//
+// @param inst : The instruction to be checked
+bool PatchReadFirstLane::isAllUsersMarkedUniform(Instruction *inst) {
+  for (auto user : inst->users()) {
+    auto userInst = dyn_cast<Instruction>(user);
+    if (m_markedUniforms.find(userInst) == m_markedUniforms.end())
+      return false;
+  }
+  return true;
+}
+
+// =====================================================================================================================
+// Try to apply readfirstlane on the given instruction
+//
+// @param inst : The instruction to be applied readfirstlane on
+// @param builder : BuildBase to use
+bool PatchReadFirstLane::applyReadFirstLane(Instruction *inst, BuilderBase &builder) {
+  // Here, we insert amdgcn_readfirstlane only if the type is float or int
+  // TODO: Support various types when backend work is ready
+  Type *uniformTy = inst->getType();
+  bool isFloatTy = uniformTy->isFloatTy();
+  if (!isFloatTy && !uniformTy->isIntegerTy(32))
+    return false;
+
```

Comment:
It would be better to check this earlier. That is, applyReadFirstLane should just check the types via an assertion, and earlier code should make sure that applyReadFirstLane is never called with types for which it doesn't apply. This helps us avoid weird inefficiencies.